### PR TITLE
[Functools] Concept Introduction Docs

### PR DESCRIPTION
**Pull request by Tarun193**

### CODE REVIEW

**Comment by Tarun193** [comment]
Can someone please review it, and I was not to able to complete this whole task as I am busy in some other work.
But if there is some some help required , just ask me out.

**Comment by BethanyG** [comment]
@Tarun193  - My sincere apologies, I got very bogged down with things and have not gotten to reviewing all the PRs I should have.  This looks good as a general concept introduction.
>  _I was not to able to complete this whole task as I am busy in some other work._

Does this mean you no longer want to work on issue #2366?  If that is the case,  I can merge this introduction, and un-assign you from the other tasks.

**Comment by Tarun193** [comment]
@BethanyG hey, sorry replying late , I just wrote the about document of the concept can you please check it.

File: concepts/functools/about.md [hunk 1]
```diff
@@ -1,2 +1,165 @@
-#TODO: Add about for this concept.
+# About
 
+The functools module is for higher-order functions: functions that act on or return other ***[functions](https://docs.python.org/3/tutorial/controlflow.html#defining-functions)***. It provides functions for working with other functions and callable objects to use or extend them without completely rewriting them.
```
Comment (BethanyG) [inline concepts/functools/about.md:3]:
```suggestion
The [functools][functools-module] module is for higher-order functions: functions that act on or return other ***[functions][defining-a-function]***. It provides functions for working with other functions and callable objects to use or extend them without completely rewriting them.

```

File: concepts/functools/about.md [hunk 2]
```diff
@@ -1,2 +1,165 @@
...
+# 10 got assigned to a because partial start assigning arguments from the left.
+
+>>> a = partial(add, b=10)
+>>> a(4)
+"got a=4, b=10"
+14
+
+# But By using the keywords we can assign the value to the arguments at right
+
+```
```
Comment (BethanyG) [inline concepts/functools/about.md:165]:
```suggestion
```

[defining-a-function]: https://docs.python.org/3/tutorial/controlflow.html#defining-functions
[functools-module]:  https://docs.python.org/3/library/functools.html
```

File: concepts/functools/about.md [hunk 3]
```diff
@@ -1,2 +1,165 @@
...
+
+## Decorators
+
+A ***[decorator](https://www.programiz.com/python-programming/decorator)*** takes in a function, adds some functionality and returns it. Decorators are example of higher order functions. We can decorate our function by writing `@decorator_function_name` just before defing function.
+
+Functions and methods are called callable as they can be called.
+
+In fact, any object which implements the special ```__call__()``` method is termed callable. So, in the most basic sense, a decorator is a callable that returns a callable.
+
+Basically, a decorator takes in a function, adds some functionality and returns it.
```
Comment (BethanyG) [inline concepts/functools/about.md:32]:
```suggestion
A ***[decorator](https://www.programiz.com/python-programming/decorator)*** takes in a function, adds some functionality and returns it.
Decorators are example of higher order functions.
We can decorate our function by writing `@decorator_function_name` just before defining function.

Functions and methods are called callable as they can be called.

In fact, any object which implements the special ```__call__()``` method is termed callable. 
So, in the most basic sense, a decorator is a callable that returns a callable.

Basically, a decorator takes in a function, adds some functionality and returns it.

```

**Comment by BethanyG** [comment]
@Tarun193  -- Thank you for getting a start on the `about.md` concept doc!   I've left a few comments.  Before I talk about the content, I just want to give you a few reminders on our [Markdown Specifications](https://exercism.org/docs/building/markdown/markdown), and our [Writing Style Guide](https://exercism.org/docs/building/markdown/style-guide).  In particular, you should be using [Markdown Reference-Style Links](https://www.markdownguide.org/basic-syntax/#reference-style-links), and making sure that [each sentence is on its own line](https://hieuphay.com/one-sentence-per-line/).

My initial comments show an example of each of these techniques.  Rather than comment on each formatting or link instance (_since these docs are going to need revisions, and re-formatting will add noise_), I think it would be cleaner if when you did your next revisions, you rework the remainder of the links and the sentences to conform to the specifications. If you could also update the `introduction.md` as well, that would be terrific. 

Moving on to the content.  I want to apologize - `decorators` (except for the use of `@functools.wraps()`) should not be part of this exercise nor part of the `about.md` for `functools`.  Decorators have their own [concept exercise and concept docs](https://github.com/exercism/python/issues/2356), which is a prerequisite to this exercise and these concepts. So your section on decorators should be re-worked to only talk about the pros and cons of using `@functools.wraps()` when writing decorators, rather than introducing decorators as a full topic, or giving examples of how to write one.   I have updated the [functools exercise spec](https://github.com/exercism/python/issues/2356) to reflect this.

Likewise, there is no need to talk about what a higher-order function is, since that will have been covered in the [Higher-Order-Functions](https://github.com/exercism/python/issues/2355) exercise.

Now for some things that _should_ be included that haven't been included yet.

For [`functools.partial()`](https://docs.python.org/3/library/functools.html#functools.partial) we' want to call out the three different "flavors" of objects produced explicitly with some prose as well as code examples.  So we want to be specific about what [`partial.func`, `partial.args`, and `partial.keywords`](https://docs.python.org/3/library/functools.html#partial-objects) **are**, give clear examples of how using `partial()` in different ways _creates_ them, and how using them **behaves** in code.  We also want to introduce [`functools.partialmethod()`](https://docs.python.org/3/library/functools.html#functools.partialmethod), and explain what the difference between the two are (one is for use with _methods_ in a class vs a stand-alone function).

Like  with `functools.partial()`and `functools.partialmethod`, we want to explain how `functools.singledispatch()` and `functools.singledispatchmethod()` are the function and method (_bound to a class_) versions of the same generic function process, and give clear examples of what that looks like in code.

Finally, we want to cover [`functools.cache()`](https://docs.python.org/3/library/functools.html#functools.cache), [`functools.lru_cache()`](https://docs.python.org/3/library/functools.html#functools.lru_cache), and  [`functools.update_wrapper`](https://docs.python.org/3/library/functools.html#functools.update_wrapper)

I have updated the spec and removed `@functools.cached_property()` (_it feels like it belongs to an exercise on class customization instead_) and `@functools.total.orderg` (_it belongs with rich comparisons_).

I am going to leave my feedback at that for now, and save more in-depth review for after you've added in the additional functions and methods.  I've also converted this to a draft PR, so that we can continue to make large revisions without sending out review notifications.  I will convert it back into a "regular" PR as we get closer to finalizing things.

Please let me know if you have any questions or issues, and apologies again for the delay in getting you feedback.  Thank you for all your hard work on this!

**Comment by Tarun193** [comment]
Sorry for not replying.
Hey I just updated the about and introduction file according to the new requirements of the concept.
@BethanyG can you please check it.

**Review by BethanyG — APPROVED** [review]
Hi @Tarun193 👋🏽 

Thanks for updating this with the new requirements.  I really appreciate it.  

Since the comments on this PR are getting quite long and noisy, I am going to approve this as-is and optimistically merge it. Should you want to continue working on it more, we can pick up discussions in a follow-on PR.  Let me know if you have any questions or issues.

**Comment by Tarun193** [comment]
Hi @BethanyG 👋
thanks for your appreciation.
No I don't want to work on it any more as the about section is good enough according to me and I don't have idea for exercise.
But if you want some some improvements feel free ask me. I will do it as per suggestions.
Thanks.

**Pull request merged**